File: public/java/038-construct-binary-tree-preorder-inorder.java
/*
 * LeetCode Problem #105: Construct Binary Tree from Preorder and Inorder Traversal
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // APPROACH 1: Recursive with Hash Map | O(n) time | O(n) space
    // EXPLAIN: First element of preorder is always the root; find it in inorder (via O(1) map lookup) to split into left/right subtrees.
    private Map<Integer, Integer> inMap;
    private int preIdx;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        inMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) inMap.put(inorder[i], i);
        preIdx = 0;
        return build(preorder, 0, inorder.length - 1);
    }
    private TreeNode build(int[] pre, int lo, int hi) {
        if (lo > hi) return null;
        TreeNode root = new TreeNode(pre[preIdx++]);
        int mid = inMap.get(root.val);
        root.left  = build(pre, lo,    mid - 1);
        root.right = build(pre, mid + 1, hi);
        return root;
    }

    // APPROACH 2: Recursive without Hash Map | O(n²) time | O(n) space
    // EXPLAIN: Same approach but finds the root in inorder with a linear scan; simpler but slower.
    public TreeNode buildTreeNoMap(int[] preorder, int[] inorder) {
        return buildNoMap(preorder, inorder, 0, preorder.length - 1, 0, inorder.length - 1);
    }
    private TreeNode buildNoMap(int[] pre, int[] in, int ps, int pe, int is, int ie) {
        if (ps > pe || is > ie) return null;
        int rootVal = pre[ps];
        TreeNode root = new TreeNode(rootVal);
        int mid = is;
        while (in[mid] != rootVal) mid++;
        int leftSize = mid - is;
        root.left  = buildNoMap(pre, in, ps + 1, ps + leftSize, is, mid - 1);
        root.right = buildNoMap(pre, in, ps + leftSize + 1, pe, mid + 1, ie);
        return root;
    }

    // APPROACH 3: Iterative with Stack | O(n) time | O(n) space
    // EXPLAIN: Process preorder elements; use an inorder pointer to decide left-child vs right-child attachment.
    public TreeNode buildTreeIterative(int[] preorder, int[] inorder) {
        if (preorder.length == 0) return null;
        TreeNode root = new TreeNode(preorder[0]);
        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        int inIdx = 0;
        for (int i = 1; i < preorder.length; i++) {
            TreeNode node = new TreeNode(preorder[i]);
            TreeNode parent = stack.peek();
            if (parent.val != inorder[inIdx]) {
                parent.left = node;
            } else {
                while (!stack.isEmpty() && stack.peek().val == inorder[inIdx]) {
                    parent = stack.pop(); inIdx++;
                }
                parent.right = node;
            }
            stack.push(node);
        }
        return root;
    }

    // APPROACH 4: Slicing (Intuitive) | O(n²) time | O(n²) space
    // EXPLAIN: Create sub-arrays for each recursive call; most readable but least efficient.
    public TreeNode buildTreeSlicing(int[] preorder, int[] inorder) {
        if (preorder.length == 0) return null;
        int rootVal = preorder[0];
        int mid = 0;
        while (inorder[mid] != rootVal) mid++;
        TreeNode root = new TreeNode(rootVal);
        root.left  = buildTreeSlicing(Arrays.copyOfRange(preorder, 1, 1 + mid),
                                       Arrays.copyOfRange(inorder, 0, mid));
        root.right = buildTreeSlicing(Arrays.copyOfRange(preorder, 1 + mid, preorder.length),
                                       Arrays.copyOfRange(inorder, mid + 1, inorder.length));
        return root;
    }

    // APPROACH 5: Standard Solution (same as Approach 1) | O(n) time | O(n) space
    // EXPLAIN: Canonical entry point — delegates to the hash-map approach.
    public TreeNode buildTreeStandard(int[] preorder, int[] inorder) {
        return buildTree(preorder, inorder);
    }
}

// Made with Bob

File: public/python/038-construct-binary-tree-preorder-inorder.py
# LeetCode Problem #105: Construct Binary Tree from Preorder and Inorder Traversal
# Difficulty: Medium
# Link: https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ─────────────────────────────────────────────
# APPROACH 1: Recursive with Hash Map | O(n) time | O(n) space
# EXPLAIN: First element of preorder is the root; find it in inorder to split left/right subtrees.
#          Use a hash map for O(1) inorder lookups and a mutable preorder index.
# WHEN: Optimal time complexity; standard interview answer.

def buildTree_hashmap(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    idx_map = {val: i for i, val in enumerate(inorder)}
    pre_idx = [0]

    def build(lo, hi):
        if lo > hi:
            return None
        root_val = preorder[pre_idx[0]]
        pre_idx[0] += 1
        root = TreeNode(root_val)
        mid = idx_map[root_val]
        root.left  = build(lo, mid - 1)
        root.right = build(mid + 1, hi)
        return root

    return build(0, len(inorder) - 1)


# ─────────────────────────────────────────────
# APPROACH 2: Recursive without Hash Map | O(n²) time | O(n) space
# EXPLAIN: Same logic but searches for root in inorder with linear scan per call.
# WHEN: Simpler code at the cost of O(n²) time; fine for small inputs.

def buildTree_no_map(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    def build(pre_s, pre_e, in_s, in_e):
        if pre_s > pre_e or in_s > in_e:
            return None
        root_val = preorder[pre_s]
        root = TreeNode(root_val)
        mid = in_s
        while inorder[mid] != root_val:
            mid += 1
        left_size = mid - in_s
        root.left  = build(pre_s + 1, pre_s + left_size, in_s, mid - 1)
        root.right = build(pre_s + left_size + 1, pre_e, mid + 1, in_e)
        return root

    return build(0, len(preorder) - 1, 0, len(inorder) - 1)


# ─────────────────────────────────────────────
# APPROACH 3: Iterative with Stack | O(n) time | O(n) space
# EXPLAIN: Process preorder elements with a stack; use the inorder pointer to decide when to switch from left-child to right-child attachment.
# WHEN: Avoids recursion; useful when call-stack depth is a concern.

def buildTree_iterative(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    if not preorder:
        return None
    root = TreeNode(preorder[0])
    stack = [root]
    in_idx = 0
    for i in range(1, len(preorder)):
        node = TreeNode(preorder[i])
        parent = stack[-1]
        if parent.val != inorder[in_idx]:
            parent.left = node
        else:
            while stack and stack[-1].val == inorder[in_idx]:
                parent = stack.pop()
                in_idx += 1
            parent.right = node
        stack.append(node)
    return root


# ─────────────────────────────────────────────
# APPROACH 4: Slicing (Intuitive) | O(n²) time | O(n²) space
# EXPLAIN: Create explicit sub-lists for each recursive call; mirrors the conceptual split most clearly.
# WHEN: Most readable but least efficient; good for understanding the algorithm.

def buildTree_slicing(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    if not preorder:
        return None
    root_val = preorder[0]
    root = TreeNode(root_val)
    mid = inorder.index(root_val)
    root.left  = buildTree_slicing(preorder[1:1 + mid], inorder[:mid])
    root.right = buildTree_slicing(preorder[1 + mid:], inorder[mid + 1:])
    return root


# ─────────────────────────────────────────────
# APPROACH 5: Standard Solution (hash map, same as Approach 1) | O(n) time | O(n) space
# EXPLAIN: Canonical version matching LeetCode's method signature.
# WHEN: Default submission entry point.

def buildTree(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    return buildTree_hashmap(preorder, inorder)


# Made with Bob
